src/devtools/client/webconsole/utils/autocomplete.ts

```diff
@@ -0,0 +1,249 @@
+      const prototypeProperties = Object.getOwnPropertyNames(Object.prototype);
+      properties.unshift(...prototypeProperties);
+    }
+
+    if (object.className === "Array") {
+      const prototypeProperties = Object.getOwnPropertyNames(Array.prototype);
+      properties.unshift(...prototypeProperties);
+    }
+  }
+
+  const prototype = preview?.prototypeValue;
+
+  // Recursively gather the properties through the prototype chain.
+  if (prototype?._object) {
+    const prototypeProperties = getPropertiesForObject(prototype._object);
+    properties.unshift(...prototypeProperties);
+  }
+
+  return properties;
+}
+
+export function getParentExpression(input: string) {
+  if (isBracketNotation(input)) {
+    return input.slice(0, input.lastIndexOf("["));
+  } else {
+    return input.slice(0, input.lastIndexOf("."));
+  }
+}
+
+export function shouldPropertyAutocomplete(input: string) {
+  if (!isAccessingObjProperty(input)) {
+    return false;
+  }
+
+  const parentExpression = getParentExpression(input);
+
+  if (!parentExpression) {
+    return false;
+  }
+
+  // Right now, this only works for direct property access (e.g., a.b)
+  // and doesn't support chained expressions (e.g., a.b.c.d). This is because
+  // we don't (yet) have the chained values in the fetched frame scope data.
+  return isIdentifier(parentExpression);
+}
+function shouldVariableAutocomplete(input: string) {
+  return isIdentifier(input);
+}
+
+function getBinding(name: string, frameScope: FrameScope) {
+  return frameScope.scope.bindings.find(b => b.name === name);
+}
+function getBindingNames(scope: FrameScope): string[] {
+  if (!scope?.scope) {
+    return [];
+  }
+
+  return scope.scope.bindings.map(b => b.name);
+}
+function getGlobalVariables(scopes: Scope) {
+  const variableNames = [];
+  const globalObject = scopes.parent.object;
+
+  if (globalObject) {
+    const properties = getPropertiesForObject(globalObject._object);
+    variableNames.push(...properties);
+  }
+
+  return variableNames;
+}
+
+export function getLastToken(input: string) {
+  if (isBracketNotation(input)) {
+    return input.slice(input.lastIndexOf("[") + 1);
+  } else {
+    return input.slice(input.lastIndexOf(".") + 1);
+  }
+}
+
+export function getCursorIndex(value: string) {
+  if (shouldPropertyAutocomplete(value)) {
+    // +1 to account for the `.` or `[`
+    return getParentExpression(value).length + 1;
+  } else {
+    return 0;
+  }
+}
+
+export function insertAutocompleteMatch(value: string, match: string) {
+  const parentExpression = getParentExpression(value);
+
+  if (shouldPropertyAutocomplete(value)) {
+    if (isBracketNotation(value)) {
+      return `${parentExpression}[${match}]`;
+    } else {
+      return `${parentExpression}.${match}`;
+    }
+  } else {
+    // For variable autocomplete.
+    return match;
+  }
+}
+
+export function getAutocompleteMatches(input: string, frameScope: FrameScope) {
+  if (!frameScope?.scope) {
```

Comment:
What does it mean when `frameScope?.scope` is falsey?

Reply:
Simplified so we can hand wave this away